File: hencoderplus/src/main/java/com/lirui/hencoderplus/MainActivity.java
package com.lirui.hencoderplus;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;

import java.util.HashMap;
import java.util.Map;

public class MainActivity extends AppCompatActivity {

    private Map<String, String> map = new HashMap<>();
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        View view = findViewById(R.id.rl_1);
        //打印出了3
        Log.i("222222222", maxDeep(view) + "");
    }
    private int maxDeep(View view) {
        //当前的view已经是最底层view了，不能往下累加层数了，返回0，代表view下面只有0层了
        if (!(view instanceof ViewGroup)) {
            return 0;
        }
        ViewGroup vp = (ViewGroup) view;
        //虽然是viewgroup，但是如果并没有任何子view，那么也已经是最底层view了，不能往下累加层数了，返回0，代表view下面只有0层了
        if (vp.getChildCount() == 0) {
            return 0;
        }
        //用来记录最大层数
        int max = 0;
        //广度遍历view
        for (int i = 0; i < vp.getChildCount(); i++) {
            //由于vp拥有子view，所以下面还有一层，因为可以+1，来叠加一层，然后再递归几岁算它的子view的层数
            int deep = maxDeep(vp.getChildAt(i)) + 1;
            //比较哪个大就记录哪个
            if (deep > max) {
                max = deep;
            }
        }
        return max;
    }
    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN) {
            map.put("1", "22");
        }
        if (event.getAction() == KeyEvent.ACTION_DOWN) {
            map.put("1", "222221");
        }
        return super.dispatchKeyEvent(event);
    }
}

File: jetpack/src/main/java/com/lirui/jetpack/MainActivity.kt
package com.lirui.jetpack

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.*
import com.lirui.jetpack.life_cycle.TestChildObserver
import com.lirui.jetpack.life_cycle.TestLifeCycleObserver
import com.lirui.jetpack.room.MyDatabase
import com.lirui.jetpack.room.Student
import com.lirui.jetpack.view_model.StudentViewModel
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class MainActivity : AppCompatActivity() {
    private val TAG = "MainActivity1"

    val scope = MainScope()
    lateinit var testLiveData:MutableLiveData<String>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
//        lifecycle.addObserver(TestChildObserver())
//
//        val model = ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory.getInstance(this.application)).get(StudentViewModel::class.java)
//        model.mLiveData?.observe(this, Observer {
//            it.forEach { student: Student ->
//                Log.d(TAG, "onCreate:${student.name}")
//            }
//        })
//        testLiveData = MutableLiveData("hahah")
//        testLiveData.observe(this, Observer {
//            sign_text.text = it
//        })
//
//        sign_text.setOnClickListener {
////            MyDatabase.getInstance(this).studentDao().insertStudent(Student("new","new","man"))
//            startActivity(Intent(this, MainActivity::class.java))
//        }
        val dao = MyDatabase.getInstance(this).studentDao()
//        dao.insertStudent(Student(1, "aaa", "2", "男"))
//        dao.insertStudent(Student(0, "bbb", "42", "女"))
//        dao.insertStudent(Student(0, "ccc", "32", "男"))
//        dao.insertStudent(Student(0, "ddd", "12", "男"))

        dao.insertStudent(Student(2, "aaa", "3", "男"))

    }

    override fun onResume() {
        super.onResume()
    }
}
